File: backend/categories.py
from fastapi import APIRouter, Depends, HTTPException, status
from typing import List
from models import CategoryCreate, CategoryResponse, CategoryStats
from database import db
from auth import get_admin_user
from bson import ObjectId
from datetime import datetime, timezone

router = APIRouter(prefix="/api/categories", tags=["Categories"])

@router.get("/stats", response_model=CategoryStats)
async def get_categories_stats(admin: dict = Depends(get_admin_user)):
    total_categories = await db.db["categories"].count_documents({})
    
    # Active categories are those used in any service
    # We use aggregate to find unique categories used in the services collection
    pipeline = [
        {"$group": {"_id": "$category"}},
        {"$count": "count"}
    ]
    cursor = db.db["services"].aggregate(pipeline)
    result = await cursor.to_list(1)
    active_categories = result[0]["count"] if result else 0
    
    return {
        "total_categories": total_categories,
        "active_categories": active_categories
    }

@router.get("", response_model=List[CategoryResponse])
async def get_categories():
    categories = await db.db["categories"].find().sort("name", 1).to_list(1000)
    return categories

@router.post("", response_model=CategoryResponse, status_code=status.HTTP_201_CREATED)
async def create_category(category: CategoryCreate, admin: dict = Depends(get_admin_user)):
    # Check if category already exists
    existing = await db.db["categories"].find_one({"name": {"$regex": f"^{category.name}$", "$options": "i"}})
    if existing:
        raise HTTPException(status_code=400, detail="Category already exists")
    
    category_dict = category.model_dump()
    category_dict["created_at"] = datetime.now(timezone.utc)
    
    result = await db.db["categories"].insert_one(category_dict)
    created = await db.db["categories"].find_one({"_id": result.inserted_id})
    return created

@router.delete("/{id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_category(id: str, admin: dict = Depends(get_admin_user)):
    if not ObjectId.is_valid(id):
        raise HTTPException(status_code=400, detail="Invalid Category ID")
        
    result = await db.db["categories"].delete_one({"_id": ObjectId(id)})
    
    if result.deleted_count == 0:
        raise HTTPException(status_code=404, detail="Category not found")
        
    return None

File: node-backend/routes/categories.js
const express = require("express");
const Category = require("../models/Category");
const Service = require("../models/Service");
const { requireAuth, requireAdmin } = require("../middleware/auth");

const router = express.Router();

router.get("/", async (req, res) => {
  try {
    const categories = await Category.find({}).sort({ name: 1 });
    return res.json(categories);
  } catch (error) {
    return res.status(500).json({ message: "Failed to fetch categories" });
  }
});

router.get("/stats", requireAuth, requireAdmin, async (req, res) => {
  try {
    const totalCategories = await Category.countDocuments();
    const activeCategoryNames = await Service.distinct("category");

    return res.json({
      total_categories: totalCategories,
      active_categories: activeCategoryNames.length
    });
  } catch (error) {
    return res.status(500).json({ message: "Failed to fetch category stats" });
  }
});

router.post("/", requireAuth, requireAdmin, async (req, res) => {
  try {
    const { name, description = "", icon = "" } = req.body;

    if (!name) {
      return res.status(400).json({ message: "Category name is required" });
    }

    const existing = await Category.findOne({ name: { $regex: `^${name.trim()}$`, $options: "i" } });
    if (existing) {
      return res.status(400).json({ message: "Category already exists" });
    }

    const category = await Category.create({
      name: name.trim(),
      description,
      icon
    });

    return res.status(201).json(category);
  } catch (error) {
    return res.status(500).json({ message: "Failed to create category" });
  }
});

router.delete("/:id", requireAuth, requireAdmin, async (req, res) => {
  try {
    const deleted = await Category.findByIdAndDelete(req.params.id);
    if (!deleted) {
      return res.status(404).json({ message: "Category not found" });
    }
    return res.status(204).send();
  } catch (error) {
    return res.status(400).json({ message: "Invalid category id" });
  }
});

module.exports = router;
